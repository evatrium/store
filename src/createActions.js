import {assign} from "@iosio/utils/lib/assign";


/**
 * Helper function for creating simple to use state control actions on the specified namespace of the store's state object.
 * @param {string} actionsName - a string to be assigned as the namespace in which to access the actions on
 * @param {Object} initialState - the state object whose property names will be used to create actions with
 * @param {Object} store - store object containing methods created by "createStore()"
 * @param {Object} actions - the global namespaced actions object
 * @returns {*} actions - the global namespaced actions object applied with the new namespace actions
 */
export const createActions = (actionsName, initialState, store, actions) => {

    /**
     * define the new namespace on the global actions object to hold the actions to be assigned
     * @type {{set: {}, get: {}, merge: merge, update: update, toggle: {}}}
     */
    actions[actionsName] = {
        set: {},
        get: {},
        toggle: {},
        update: () => {
        },
        merge: () => {
        }
    };

    /**
     * for updating multiple state properties at a time
     * must assign/spread the existing state that is passed to the callback function if updating only a few properties.
     * otherwise, the properties that are not returned that do exist on the previous state will be missing on the next state update
     *
     * @example
     *  myNameSpacedActions.update(state=>({
     *      ...state,//all the other existing properties
     *      someExistingProp: 'some new value'
     *  }));
     *
     * @param {function} cb - a callback function providing the previous state as an argument
     * @returns {undefined} - returns nothing
     */
    actions[actionsName].update = (cb) => {

        let state = store.getState();
        //just to be sure not to modify the original reference, make a copy
        let copyOfExistingState = assign({}, state);
        //get the object on the namespace which to make the modifications on
        let oldNameSpaceState = state[actionsName];
        //provide the state to the callback and retrieve the results
        let updatedNameSpaceState = cb(oldNameSpaceState);
        //take the results and apply them on a namespace property of a new object
        let toBeMerged = assign({}, {[actionsName]: updatedNameSpaceState});
        //merge the update to the state copy
        let allState = assign(copyOfExistingState, toBeMerged);
        //set the state on the store
        store.setState(allState);

    };

    /**
     * merges new properties on the state namespace
     * but without the need to re-spread the previous state on the return object
     * @example
     *  myNameSpacedActions.merge({
     *          someExistingProp: 'some new value',
     *          anotherExistingProp: 'another new value'
     *      });
     *
     * @param {Object} propsToMergeToNameSpace - new object containing properties to merge onto the namespaced state
     * @returns {undefined} - returns nothing
     */
    actions[actionsName].merge = (propsToMergeToNameSpace) => {

        let state = store.getState();

        let copyOfExistingState = assign({}, state);

        let oldNameSpaceState = state[actionsName];

        let updatedNameSpaceState = assign(oldNameSpaceState, propsToMergeToNameSpace);

        let toBeMerged = {[actionsName]: updatedNameSpaceState};

        let allState = assign(copyOfExistingState, toBeMerged);

        store.setState(allState);
    };

    /**
     * gets the state object that belongs only to the namespace
     * @returns {Object} - namespace state
     */
    actions[actionsName].getState = () => {
        const namespaceState = store.getState()[actionsName];
        return assign({}, namespaceState);
    }


    /*
        for every property on the initialState object
        create a set, get, and toggle function
     */

    for (let key in initialState) {

        /**
         * sets a single property value on the state
         * @param {*} newValue - value to set
         * @returns {undefined} - returns nothing
         */
        actions[actionsName].set[key] = (newValue) => {

            let state = store.getState();

            let copyOfExistingState = assign({}, state);

            let oldNameSpaceState = state[actionsName];

            let propsToMergeToNameSpace = {[key]: newValue};

            let updatedNameSpaceState = assign(oldNameSpaceState, propsToMergeToNameSpace);

            let toBeMerged = {[actionsName]: updatedNameSpaceState};

            let allState = assign(copyOfExistingState, toBeMerged);

            store.setState(allState);
        };

        /**
         * gets a specific property value on the state
         * @returns {*}
         */
        actions[actionsName].get[key] = () => {

            let nameSpaceKeyValue = store.getState()[actionsName][key];

            let tobBeCopied = {[key]: nameSpaceKeyValue};

            let copied = assign({}, tobBeCopied);

            return copied[key];
        };

        /**
         * gets the previous value and updates it with the inverse
         */
        actions[actionsName].toggle[key] = () => {
            actions[actionsName].set[key](!actions[actionsName].get[key]())
        };
    }
    return actions;
};