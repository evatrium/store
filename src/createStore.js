import {assign} from "@iosio/utils/lib/assign";

/**
 * creates a store to store state, get and set state, attach new state properties and subscribe to updates
 * @param {undefined|Object} state - the state object to hold all the namespaced state
 * @returns {{attachState: attachState, getState: (function(): *), unsubscribe: unsubscribe, subscribe: (function(*=): function(): void), setState: setState}}
 */
export const createStore = (state) => {
    /**
     * holds all teh subscription callback functions
     * @type {Array}
     */
    let listeners = [];

    /**
     * state object
     * @type {Object|{}} - holds state
     */
    state = state || {};


    /**
     * removes the listener
     * @param {function} listener - named callback function to be removed from listeners when unsubscribed
     * @returns {undefined} - returns nothing
     */
    const unsubscribe = (listener) => {
        let out = [];
        for (let i = 0; i < listeners.length; i++) {
            if (listeners[i] === listener) {
                listener = null;
            } else {
                out.push(listeners[i]);
            }
        }
        listeners = out;
    };

    return {
        unsubscribe,
        /**
         * gets the entire state object
         * @returns {Object} - the state object
         */
        getState: () => state,
        /**
         * merges new state to the existing state object unless overwrite is true, thus replaces the state.
         * triggers a state update, calling all listeners, passing the updated state to the callbacks
         * @param {Object} update - to be merged to state
         * @param {boolean} overwrite - replaces the existing state with the update if true
         * @returns {undefined} - returns nothing
         */
        setState: (update, overwrite) => {
            state = overwrite ? update : assign(assign({}, state), update);
            let current = listeners;
            for (let i = 0; i < current.length; i++) current[i](state);
        },
        /**
         * pass a callback function to subscribe to state updates
         * @param {function} listener - callback function
         * @returns {function(): *} - call this function to unsubscribe this specific listener
         */
        subscribe: (listener) => {
            listeners.push(listener);
            return () => unsubscribe(listener);
        },
        /**
         * attaches new state to the state object without triggering an update
         * @param {object} new_state - stage object to be merged to existing state
         * @returns {undefined} - returns nothing
         */
        attachState: (new_state) => {
            state = assign(assign({}, state), new_state);
        },
    }
};



