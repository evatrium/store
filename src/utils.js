/**
 * provides the convenience for optionally using comma separated values to pluck the properties off of the namespace
 * can be used for mapState and mapLogic
 * @example
 *
 * mapState: {
 *      access: 'loggedIn,user'
 * }
 *
 * //equal to doing
 * mapState: ({access})=>({
 *      loggedIn: access.loggedIn,
 *      user: access.user
 * })
 *
 * @param {function|Object} mapper - function or object containing the values to map
 * @param {Object} source
 * @returns {Object} - object containing the plucked properties
 */
export const mapIt = (mapper, source) => {

    if (typeof mapper === 'function') {

        return mapper(source);

    } else if (typeof mapper === 'object') {

        let selected = {};

        let keys = Object.keys(mapper);

        for (let i = 0; i < keys.length; i++) {

            const namespace = keys[i];

            if (!source[namespace]) {
                console.error(`"${namespace}" does not exist or the namespace has not been attached to state yet`);
                break;
            }

            let props = mapper[namespace];

            if (typeof props === 'string') props = props.split(/\s*,\s*/);

            for (let i = 0; i < props.length; i++) {
                selected[props[i]] = source[namespace][props[i]];
            }
        }

        return selected;

    } else {

        return {};
    }
};