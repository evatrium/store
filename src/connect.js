import React, {Component, createElement, createContext} from "react";
import {assignAll} from '@iosio/utils/lib/assignAll';
import {mapIt} from "./utils";

const StoreContext = createContext();
export const Provider = ({store, actions, collective, children}) =>
    <StoreContext.Provider value={{store, actions, collective}}>{children} </StoreContext.Provider>;

const getDisplayName = (WrappedComponent) => WrappedComponent.displayName || WrappedComponent.name || 'Component';

export const connect = ({mapState, mapActions, mapLogic, additionalProps}) => Child => {
    function Wrapper(props, context) {
        Component.call(this, props, context);
        const {store, actions, collective} = context;
        let mappedProps = {},
            state = {},
            update = () => {
            }, sub, unsub;

        if (actions && mapActions) {
            mappedProps = mapIt(mapActions, actions.getActions());
        }
        if (collective && mapLogic) {
            mappedProps = assignAll(mappedProps, [mapIt(mapLogic, collective.getCollective())])
        }
        if (store && mapState) {
            state = mapIt(mapState, store.getState());
            update = () => {
                if (this.mounted) {
                    let mapped = mapIt(mapState, store.getState());
                    for (let i in mapped) {
                        if (mapped[i] !== state[i]) {
                            state = mapped;
                            this.mounted && this.forceUpdate();
                            return;
                        }
                    }
                    for (let i in state) {
                        if (!(i in mapped)) {
                            state = mapped;
                            this.mounted && this.forceUpdate();
                            return
                        }
                    }
                }
            };
            sub = () => store.subscribe(update);
            unsub = () => store.unsubscribe(update);
        }

        this.componentDidMount = () => {
            this.mounted = true;
            sub && sub();
        };
        this.componentWillUnmount = () => {
            this.mounted = false;
            unsub && unsub();
        };
        this.render = () =>
            createElement(Child, assignAll({}, [mappedProps, this.props, state, additionalProps]));
    }

    const TheComponent = (Wrapper.prototype = new Component()).constructor = Wrapper;
    TheComponent.displayName = `${getDisplayName(Child)}_connected`;
    TheComponent.contextType = StoreContext;
    return TheComponent;
};