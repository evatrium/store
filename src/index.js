import {createStore} from "./createStore";
import {createActions} from "./createActions";
import {Provider, connect} from "./connect";

export {createStore, createActions, Provider, connect};
export default createStore;
//40.57