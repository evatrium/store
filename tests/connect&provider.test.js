import {createStore, createActions, connect, Provider} from "../src";
import React from 'react';
import renderer from 'react-test-renderer';

const namespace = 'namespace';


const SetUp = () => {
    const store = createStore();

    let actions = {};
    const getActions = () => actions;
    let collective = {
        [namespace]:{
            testFunc: () => {
                return "hello"
            }
        }

    };
    const getCollective = () => collective;

    const initialState = {count: 0, name: 'Tom'};

    store.attachState({[namespace]: initialState});

    createActions(namespace, initialState, store, actions);

    const Prov = ({children}) => (
        <Provider store={store} actions={{getActions}} collective={{getCollective}}>{children}</Provider>
    );


    return {
        Prov,
        ...actions[namespace]
    }
};

describe('Provider provides state updates to connect', () => {


    it('renders without crashing', () => {

        const {Prov} = SetUp();

        const tree = renderer.create(
            <Prov>
                <div>test</div>
            </Prov>
        );

        expect(tree).toMatchSnapshot();
    });

    it('passes the right props to the component', () => {

        const {Prov} = SetUp();

        const MyLabel = connect({mapState: {[namespace]: 'name'}})((props) => {
            expect(props.name).toBe('Tom');
            return <div>{props.name}</div>
        });


        const tree = renderer.create(
            <Prov>
                <MyLabel/>
            </Prov>
        );

        expect(tree).toMatchSnapshot();

    });

    it('updates to state are reflected on the component', () => {

        const {Prov, set} = SetUp();

        set.name('Alpha')

        const MyLabel2 = connect({mapState: {[namespace]: 'name'}})((props) => {
            expect(props.name).toBe('Alpha');
            return <div>{props.name}</div>
        });


       let tree = renderer.create(
            <Prov>
                <MyLabel2/>
            </Prov>
        );

        expect(tree).toMatchSnapshot();

    });



    it('maps actions to props', () => {

        const {Prov} = SetUp();


        const MyLabel = connect({mapActions: {[namespace]: 'set,get'}})((props) => {

            expect(props).toMatchObject({
                set: {
                    count: expect.any(Function),
                    name:expect.any(Function)
                },
                get:{
                    count: expect.any(Function),
                    name:expect.any(Function)
                },
            });

            return <div>mapped actions: {JSON.stringify(props, null, 4)}</div>
        });


        let tree = renderer.create(
            <Prov>
                <MyLabel/>
            </Prov>
        );

        expect(tree).toMatchSnapshot();

    });

    it('maps logic to props', () => {

        const {Prov} = SetUp();


        const MyLabel = connect({mapLogic: {[namespace]: 'testFunc'}})((props) => {

            expect(props.testFunc).toBeInstanceOf(Function);

            return <div>mapped logic: {JSON.stringify(props, null, 4)}</div>
        });


        let tree = renderer.create(
            <Prov>
                <MyLabel/>
            </Prov>
        );

        expect(tree).toMatchSnapshot();

    });

});