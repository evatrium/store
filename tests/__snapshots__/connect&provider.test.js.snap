// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider provides state updates to connect maps actions to props 1`] = `
Array [
  <div>
    mapped actions: 
    {
    "set": {},
    "get": {}
}
  </div>,
  " ",
]
`;

exports[`Provider provides state updates to connect maps logic to props 1`] = `
Array [
  <div>
    mapped logic: 
    {}
  </div>,
  " ",
]
`;

exports[`Provider provides state updates to connect passes the right props to the component 1`] = `
Array [
  <div>
    Tom
  </div>,
  " ",
]
`;

exports[`Provider provides state updates to connect renders without crashing 1`] = `
Array [
  <div>
    test
  </div>,
  " ",
]
`;

exports[`Provider provides state updates to connect updates to state are reflected on the component 1`] = `
Array [
  <div>
    Alpha
  </div>,
  " ",
]
`;
